File: day16/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	lineRegex := regexp.MustCompile(`Valve (\w+) has flow rate=(\d+); tunnels? leads? to valves? (.*)`)

	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var valves []Valve
	var nodes []string
	var edges []Edge[string]

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		matches := lineRegex.FindStringSubmatch(scanner.Text())
		if len(matches) == 0 {
			log.Fatalln("Invalid line")
		}
		name := matches[1]
		rate, err := strconv.Atoi(matches[2])
		if err != nil {
			log.Fatalln(err)
		}
		nodes = append(nodes, name)
		if rate > 0 {
			valves = append(valves, Valve{
				Name: name,
				Rate: rate,
			})
		}
		for _, i := range strings.Split(matches[3], ", ") {
			edges = append(edges, Edge[string]{
				A: name,
				B: i,
				C: 1,
			})
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	sort.Slice(valves, func(i, j int) bool {
		return valves[i].Rate > valves[j].Rate
	})
	for n, i := range valves {
		k := i
		k.ID = 1 << n
		valves[n] = k
	}

	dist := FloydWarshall(nodes, edges)

	fmt.Println("Part 1:", searchMax(0, 30, "AA", map[string]struct{}{}, valves, dist, 0))

	pathMap := map[int]int{}
	searchPaths(0, 0, 26, "AA", map[string]struct{}{}, pathMap, valves, dist)
	allPaths := make([]Path, 0, len(pathMap))
	for k, v := range pathMap {
		allPaths = append(allPaths, Path{
			ID:   k,
			Flow: v,
		})
	}
	sort.Slice(allPaths, func(i, j int) bool {
		return allPaths[i].Flow > allPaths[j].Flow
	})

	maxFlow := 0
	for n, i := range allPaths {
		if i.Flow*2 <= maxFlow {
			break
		}
		for _, j := range allPaths[n+1:] {
			flow := i.Flow + j.Flow
			if flow <= maxFlow {
				break
			}
			if i.ID&j.ID != 0 {
				continue
			}
			if flow > maxFlow {
				maxFlow = flow
			}
		}
	}
	fmt.Println("Part 2:", maxFlow)
}

type (
	Path struct {
		ID   int
		Flow int
	}

	Valve struct {
		Name string
		ID   int
		Rate int
	}
)

func searchMax(acc int, remaining int, pos string, toggled map[string]struct{}, valves []Valve, dist PairwiseDistances[string], candidate int) int {
	if remaining <= 0 {
		return acc
	}

	{
		bound := 0
		t := remaining
		for _, i := range valves {
			if t <= 2 {
				break
			}
			if _, ok := toggled[i.Name]; ok {
				continue
			}
			t -= 2
			bound += t * i.Rate
		}
		if acc+bound < candidate {
			return 0
		}
	}

	maxFlow := acc
	if maxFlow > candidate {
		candidate = maxFlow
	}

	for _, i := range valves {
		if _, ok := toggled[i.Name]; ok {
			continue
		}
		cost, ok := dist.EdgeCost(pos, i.Name)
		if !ok {
			continue
		}
		nextRemaining := remaining - cost - 1
		if nextRemaining <= 0 {
			continue
		}
		toggled[i.Name] = struct{}{}
		flow := searchMax(acc+i.Rate*nextRemaining, nextRemaining, i.Name, toggled, valves, dist, candidate)
		delete(toggled, i.Name)
		if flow > maxFlow {
			maxFlow = flow
			if maxFlow > candidate {
				candidate = maxFlow
			}
		}
	}

	return maxFlow
}

func searchPaths(acc int, curPath int, remaining int, pos string, toggled map[string]struct{}, allPaths map[int]int, valves []Valve, dist PairwiseDistances[string]) {
	if remaining <= 0 {
		return
	}

	if v, ok := allPaths[curPath]; !ok || acc > v {
		allPaths[curPath] = acc
	}

	for _, i := range valves {
		if _, ok := toggled[i.Name]; ok {
			continue
		}
		cost, ok := dist.EdgeCost(pos, i.Name)
		if !ok {
			continue
		}
		nextRemaining := remaining - cost - 1
		if nextRemaining <= 0 {
			continue
		}
		toggled[i.Name] = struct{}{}
		searchPaths(acc+i.Rate*nextRemaining, curPath|i.ID, nextRemaining, i.Name, toggled, allPaths, valves, dist)
		delete(toggled, i.Name)
	}
}

File: day16/src/main.rs
use regex::Regex;
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

mod floydwarshall;

const PUZZLEINPUT: &str = "input.txt";

type BoxError = Box<dyn std::error::Error>;
type BoxResult<T> = Result<T, BoxError>;

fn main() -> BoxResult<()> {
    let line_regex =
        Regex::new(r"Valve (\w+) has flow rate=(\d+); tunnels? leads? to valves? (.*)").unwrap();

    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut valves = Vec::new();
    let mut nodes = Vec::new();
    let mut edges = Vec::new();

    for line in reader.lines() {
        let line = line?;
        let captures = line_regex.captures(&line).ok_or("Invalid line")?;
        let name = captures.get(1).ok_or("Invalid line")?.as_str().to_string();
        let rate = captures
            .get(2)
            .ok_or("Invalid line")?
            .as_str()
            .parse::<i32>()?;
        nodes.push(name.clone());
        if rate > 0 {
            valves.push(Valve {
                name: name.clone(),
                id: 0,
                rate,
            })
        }
        for i in captures.get(3).ok_or("Invalid line")?.as_str().split(", ") {
            edges.push(floydwarshall::Edge {
                a: name.clone(),
                b: i.to_string(),
                c: 1,
            });
        }
    }

    valves.sort_by(|a, b| b.rate.cmp(&a.rate));
    for (n, i) in valves.iter_mut().enumerate() {
        i.id = 1 << n;
    }

    let dist = floydwarshall::compute(&nodes[..], &edges[..]);
    println!(
        "Part 1: {}",
        search_max(0, 30, "AA", &mut HashSet::new(), &valves, &dist, 0)
    );

    let mut path_map = HashMap::new();
    search_paths(
        0,
        0,
        26,
        "AA",
        &mut HashSet::new(),
        &mut path_map,
        &valves,
        &dist,
    );
    let mut all_paths = path_map
        .into_iter()
        .map(|(id, flow)| Path { id, flow })
        .collect::<Vec<_>>();
    all_paths.sort_by(|a, b| b.flow.cmp(&a.flow));

    let mut max_flow = 0;
    for (n, i) in all_paths.iter().enumerate() {
        if i.flow * 2 <= max_flow {
            break;
        }
        for j in &all_paths[n + 1..] {
            let flow = i.flow + j.flow;
            if flow <= max_flow {
                break;
            }
            if i.id & j.id != 0 {
                continue;
            }
            if flow > max_flow {
                max_flow = flow;
            }
        }
    }
    println!("Part 2: {}", max_flow);

    Ok(())
}

struct Path {
    id: u32,
    flow: i32,
}

struct Valve {
    name: String,
    id: u32,
    rate: i32,
}

fn search_max(
    acc: i32,
    remaining: i32,
    pos: &str,
    toggled: &mut HashSet<String>,
    valves: &[Valve],
    dist: &floydwarshall::PairwiseDistances<String>,
    mut candidate: i32,
) -> i32 {
    if remaining <= 0 {
        return acc;
    }
    {
        let mut bound = 0;
        let mut t = remaining;
        for i in valves {
            if t <= 2 {
                break;
            }
            if toggled.contains(&i.name) {
                continue;
            }
            t -= 2;
            bound += t * i.rate;
        }
        if acc + bound < candidate {
            return 0;
        }
    }

    let mut max_flow = acc;
    if max_flow > candidate {
        candidate = max_flow;
    }

    for i in valves {
        if toggled.contains(&i.name) {
            continue;
        }
        let cost = match dist.edge_cost(&pos.to_string(), &i.name) {
            Some(c) => c,
            None => continue,
        };
        let next_remaining = remaining - cost - 1;
        if next_remaining <= 0 {
            continue;
        }
        toggled.insert(i.name.clone());
        let flow = search_max(
            acc + i.rate * next_remaining,
            next_remaining,
            &i.name,
            toggled,
            valves,
            dist,
            candidate,
        );
        toggled.remove(&i.name);
        if flow > max_flow {
            max_flow = flow;
            if max_flow > candidate {
                candidate = max_flow;
            }
        }
    }

    max_flow
}

fn search_paths(
    acc: i32,
    cur_path: u32,
    remaining: i32,
    pos: &str,
    toggled: &mut HashSet<String>,
    all_paths: &mut HashMap<u32, i32>,
    valves: &[Valve],
    dist: &floydwarshall::PairwiseDistances<String>,
) {
    if remaining <= 0 {
        return;
    }
    if match all_paths.get(&cur_path) {
        Some(&v) => acc > v,
        None => true,
    } {
        all_paths.insert(cur_path, acc);
    }

    for i in valves {
        if toggled.contains(&i.name) {
            continue;
        }
        let cost = match dist.edge_cost(&pos.to_string(), &i.name) {
            Some(c) => c,
            None => continue,
        };
        let next_remaining = remaining - cost - 1;
        if next_remaining <= 0 {
            continue;
        }
        toggled.insert(i.name.clone());
        search_paths(
            acc + i.rate * next_remaining,
            cur_path | i.id,
            next_remaining,
            &i.name,
            toggled,
            all_paths,
            valves,
            dist,
        );
        toggled.remove(&i.name);
    }
}
